hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/JoinHashPhysicalRel.java

```diff
@@ -58,9 +58,30 @@ public Join copy(
         return new JoinHashPhysicalRel(getCluster(), traitSet, left, right, conditionExpr, joinType);
     }
 
+    /**
+     * Cost calculation of Hash Join relation. It does not rely on children cost.
+     * <p>
+     * Hash Join algorithm is more advanced join algorithm, where it builds a hash table for left row set, and then
+     * compare each row from the right side
+     * Speaking of cost estimation, we are accounting the next properties:
+     * - produced row count is estimated as L * R, with assumption that the JOIN predicate has maximum selectivity.
+     * - processed row count is estimated as L + R, because we traverse both sides once per JOIN.
+     * - CPU cost estimation is a processed row multiplied by cost to build a hash table, and left and right rows comparison.
+     * <p>
+     * The perfect assumption also must include memory (what is important in case of hash table) and IO cost estimation,
+     * as well as a selectivity for a right row set.
+     */
     @Override
     @Nullable
     public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
-        return super.computeSelfCost(planner, mq).multiplyBy(COST_FACTOR);
+        double leftRowCount = mq.getRowCount(getLeft());
+        double rightRowCount = mq.getRowCount(getRight());
+        // TODO: introduce selectivity estimator, but ATM we taking the worst case scenario : selectivity = 1.0.
+        double producedRowCount = leftRowCount * /* TODO: selectivity */ rightRowCount;
+        double processedRowCount = leftRowCount + /* TODO: selectivity */ rightRowCount;
+
+        double cpu = Cost.HASH_JOIN_MULTIPLIER * processedRowCount * Cost.JOIN_ROW_CMP_MULTIPLIER;
```

Comment:
Since _hash join build cost_ covers element comparisons, we should only multiply _R_ with _row comparison cost_.
```suggestion
        double cpu = leftRowCount * Cost.HASH_JOIN_MULTIPLIER + rightRowCount * Cost.JOIN_ROW_CMP_MULTIPLIER;
```

Reply:
True, it's a more accurate estimation and my typo.